server/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -79,91 +79,52 @@ record LoadedPlugin(PluginInfo info, Plugin instance, ClassLoader loader) {}
-     * @param settings The settings of the system
+     *
+     * @param settings         The settings of the system
      * @param modulesDirectory The directory modules exist in, or null if modules should not be loaded from the filesystem
      * @param pluginsDirectory The directory plugins exist in, or null if plugins should not be loaded from the filesystem
-     * @param classpathPlugins Plugins that exist in the classpath which should be loaded
      */
-    public PluginsService(
-        Settings settings,
-        Path configPath,
-        Path modulesDirectory,
-        Path pluginsDirectory,
-        Collection<Class<? extends Plugin>> classpathPlugins
-    ) {
+    public PluginsService(Settings settings, Path configPath, Path modulesDirectory, Path pluginsDirectory) {
         this.settings = settings;
         this.configPath = configPath;
 
-        List<LoadedPlugin> pluginsLoaded = new ArrayList<>();
-        List<PluginInfo> pluginsList = new ArrayList<>();
-        // we need to build a List of plugins for checking mandatory plugins
-        final List<String> pluginsNames = new ArrayList<>();
-
-        // first we load plugins that are on the classpath. this is for tests
-        for (Class<? extends Plugin> pluginClass : classpathPlugins) {
-            Plugin plugin = loadPlugin(pluginClass, settings, configPath);
-            PluginInfo pluginInfo = new PluginInfo(
-                pluginClass.getName(),
-                "classpath plugin",
-                "NA",
-                Version.CURRENT,
-                "1.8",
-                pluginClass.getName(),
-                null,
-                Collections.emptyList(),
-                false,
-                PluginType.ISOLATED,
-                "",
-                false
-            );
-            if (logger.isTraceEnabled()) {
-                logger.trace("plugin loaded from classpath [{}]", pluginInfo);
-            }
-            pluginsLoaded.add(new LoadedPlugin(pluginInfo, plugin, null));
-            pluginsList.add(pluginInfo);
-            pluginsNames.add(pluginInfo.getName());
-        }
-
         Set<PluginBundle> seenBundles = new LinkedHashSet<>();
         List<PluginInfo> modulesList = new ArrayList<>();
+
         // load modules
         if (modulesDirectory != null) {
             try {
                 Set<PluginBundle> modules = PluginsUtils.getModuleBundles(modulesDirectory);
-                for (PluginBundle bundle : modules) {
-                    modulesList.add(bundle.plugin);
-                }
+                modules.stream().map(b -> b.plugin).forEach(modulesList::add);
                 seenBundles.addAll(modules);
             } catch (IOException ex) {
                 throw new IllegalStateException("Unable to initialize modules", ex);
             }
         }
 
-        // now, find all the ones that are in plugins/
+        // load plugins
+        List<PluginInfo> pluginsList = new ArrayList<>();
```

Comment:
Does it make sense for us to switch this to a Set instead of a list? It seems that PluginInfo has proper equals and hashcode, although the equals seems to also check version at the moment? It might be good to figure out if the comment on equals in PluginInfo should taken care of now as well. Should we be allowing multiple versions of equally named plugins? Same for modules, I think we shouldn't be allowing duplicates, perhaps even error out if we find some?

Reply:
Looks like Version got added to PluginInfo a long time ago: abf9a866788b2fab855b50728670e2aa2dd42c5a

I don't see any reason in that commit for it to be there. I'll put out a quick fix for it on a different PR in case there's some sneaky dependence on it that we'd catch in tests.

We catch duplicate-named plugins and modules in our plugin-loading logic a little ways down (triggered by `loadBundles(seenBundles)`, and I added a test to catch this in https://github.com/elastic/elasticsearch/pull/86891/files#diff-099587538f96b160a9033106117540c99ffb60e00f8e4dfe8f54a2091a262943 , so I think for the moment we're good there.

Reply:
Ah, PluginInfo's equality check was briefly commented on in PR review, but not really discussed: https://github.com/elastic/elasticsearch/pull/4378/files#r9292810